framework/src/dirackernels/DiracKernel.C

```diff
@@ -10,66 +10,40 @@
-  params += BlockRestrictable::validParams();
-
-  params.addParam<bool>("use_displaced_mesh",
-                        false,
-                        "Whether or not this object should use the displaced mesh for computation. "
-                        "Note that in the case this is true but no displacements are provided in "
-                        "the Mesh block the undisplaced mesh will still be used.");
-
-  params.addParam<bool>(
-      "drop_duplicate_points",
-      true,
-      "By default points added to a DiracKernel are dropped if a point at the same location"
-      "has been added before. If this option is set to false duplicate points are retained"
-      "and contribute to residual and Jacobian.");
-
-  MooseEnum point_not_found_behavior("ERROR WARNING IGNORE", "IGNORE");
-  params.addParam<MooseEnum>(
-      "point_not_found_behavior",
-      point_not_found_behavior,
-      "By default (IGNORE), it is ignored if an added point cannot be located in the "
-      "specified subdomains. If this option is set to ERROR, this situation will result in an "
-      "error. If this option is set to WARNING, then a warning will be issued.");
-
-  params.addParamNamesToGroup("use_displaced_mesh drop_duplicate_points", "Advanced");
-
-  params.registerBase("DiracKernel");
-
+  InputParameters params = DiracKernelBase::validParams();
+  if (std::is_same<T, Real>::value)
+    params.registerBase("DiracKernel");
+  else if (std::is_same<T, RealVectorValue>::value)
+    params.registerBase("VectorDiracKernel");
+  else
+    ::mooseError("unsupported DiracKernelTempl specialization");
   return params;
 }
 
-DiracKernel::DiracKernel(const InputParameters & parameters)
-  : ResidualObject(parameters),
-    CoupleableMooseVariableDependencyIntermediateInterface(this, false),
-    MooseVariableInterface<Real>(this,
-                                 false,
-                                 "variable",
-                                 Moose::VarKindType::VAR_NONLINEAR,
-                                 Moose::VarFieldType::VAR_FIELD_STANDARD),
-    MaterialPropertyInterface(this, Moose::EMPTY_BLOCK_IDS, Moose::EMPTY_BOUNDARY_IDS),
-    GeometricSearchInterface(this),
-    BlockRestrictable(this),
-    _var(mooseVariableField()),
-    _coord_sys(_assembly.coordSystem()),
-    _dirac_kernel_info(_subproblem.diracKernelInfo()),
+template <typename T>
+DiracKernelTempl<T>::DiracKernelTempl(const InputParameters & parameters)
+  : DiracKernelBase(parameters),
+    MooseVariableInterface<T>(this,
+                              false,
+                              "variable",
+                              Moose::VarKindType::VAR_NONLINEAR,
+                              std::is_same<T, Real>::value ? Moose::VarFieldType::VAR_FIELD_STANDARD
+                                                           : Moose::VarFieldType::VAR_FIELD_VECTOR),
+    _var(this->mooseVariableField()),
     _current_elem(_var.currentElem()),
```

Comment:
You can move this to the base class and populate with `_assembly.elem()` instead of `_var.currentElem()` and then you can move the `currentPointCachedID` method into the base class